### DISCUSSION TITLE
Clean seperation of "internal" and "external" field names

### DISCUSSION THREAD

**Discussion by Rawk** [pydantic/pydantic#2889] [Feature Request]
Is there an outspoken idea for separation of internal (deserialized) and external (serialized) field names?

What is the correct way of handling field names not representable as a Python variable?

A combination of `Field.alias` and `.json/.dict(by_alias=True)` can be used to simulate this. But if you forget to use `by_alias=True` the serialization is invalid. Using alias also have another semantic meaning: "This value can come in this _additional_ property name".

I propose something like this:

- External field names are by default equal to the **internal** field name.
- Always load data using only **external** field names or aliases.
- Work with the deserialized object in Python using only **internal** field names.
- Dump by default uses **external** field names. (.json() and .dict())
- Schemas by default uses **external** field names. (.schema() and .schema_json())

Example:

```python
class M(BaseModel)
  class_name: str = Field(..., name="class")  # internal name = "class_name", external name = "class"

m = M.parse_raw('{"class": "2B"}')   # Load using only the external name or aliases
assert m.class_name == "2B"      # Work with the object using only the internal name
assert m.json() == '{"class": "2B"}' # Dump by default uses the external name
assert m.dict() == {"class": "2B"}   # Dump by default uses the external name
m2 = M(class_name="2B")          # __init__ uses only internal name
```

I think this would cover common use cases in a clean, expected, way.

Some use cases:
- External format uses property names not suitable for Python variable name.
- External format uses short cryptic names (ex. to save bytes). Use longer names internally to make code more readable.
- External format uses UpperCamelCase but snake_case fits nicer in Python. (Mabye a Field.name_generator could be used.)
- Pass the object or model to someone elses code (for example as dependency injection), and expect it to dump/load with external names.

**Other libraries**
The Rust library Serde uses field attribute `#[serde(rename = "class")]`, and can optionally specify different rename for serialization and deserialization.

Google Gson library for Java uses the field decorator `@SerializedName("class")`.

JSON.NET in C# uses field property decorator `[JsonProperty("class")]`.

**Comment by pawamoy** [comment]
Very good proposal in my opinion, I'm a bit sad to see it was proposed almost one year ago and never had any response 😕 
Update: well, the current implementation allows it and is explicit, which is not so bad I think. It took me some time to piece everything together though.

```python
class PayloadAttachment(BaseModel):
    class Config:
        allow_population_by_field_name = True

    guid: str = Field(alias="PJ_Guid")
    gnf_or_sae: DocumentBackend = Field(DocumentBackend.GNF, alias="PJ_GNFOuSAE")


# ...

response = await client.post(path, json=payload.dict(by_alias=True, exclude_unset=True))
```

With this, you serialiaze/deserialize to/from aliases instead of internal names, and you can still build instances using internal names. That's quite nice.

**Reply by Rawk** [reply]
Thanks for you comment. I agree that it is not so bad, but it is not great either.

As i see it, there is multiple problems in your solution.
1. Internal field names "gnf_or_sae" and "guid" is also allowed when loading. It is often fine, but not strictly correct, and sometimes a bug.
2. There can be multiple aliases. [Which one is used?](https://pydantic-docs.helpmanual.io/usage/model_config/#alias-precedence) This adds complexity.
3. Somewhere, someone might forget to add `by_alias=True`. Maybe it even happens to work for a while, because the other end uses the same internal names (see 1.).
4. Say you have a function: `def create_response(model: BaseModel, status, headers, ...) -> bytes`. That function now also must take and forward `by_alias: bool`. (and callers must remember use it, see 3.).
5. `Config.allow_population_by_field_name` is extra complexity not needed.
6. As you wrote: "It took me some time to piece everything together though".

Your example would become:
```python
class PayloadAttachment(BaseModel):
    guid: str = Field(name="PJ_Guid")
    gnf_or_sae: DocumentBackend = Field(DocumentBackend.GNF, name="PJ_GNFOuSAE")

# ...

response = await client.post(path, json=payload.dict(exclude_unset=True))
```

The problem is that `alias` was solving a different problem: Allowing additional names for a field when *loading*. This is good!
But, it has then come to also be used for fixing problems that arise from the missing separation of internal and external field names.

But this can probably not be fixed without a breaking change, or adding additional complexity and confusion by adding this while keeping alias as it is. So i don't know what i want really.. :-) Something for Pydantic 2.0 maybe.

**Reply by pawamoy** [reply]
I see, thanks for the detailed answer, that's quite interesting. Yeah, it's probably something Samuel and other maintainers could consider for some future version 🙂

**Comment by Ed1123** [comment]
This certainly will be a terrific addition. My problem is that my company's API fields are "interested".

Searching for a solution for my problem using Pydantic I came to `allow_population_by_field_name = True` and `alias=`. Another problem they have is that Pylance doesn't recognize it. If I want to use it, I must add multiple `# type: ignore`.

I also encounter an implementation of Rust serde in python (https://github.com/rossmacarthur/serde). It even has the rename attribute for the fields. The only problem is that the repository is no longer maintained (maybe taking the code and merging it to Pydantic?)

I hope this comes to reality someday. 😄  If there are plans to add it, I'll be happy to help.

**Comment by alexanderankin** [comment]
several people are saying this:
https://github.com/pydantic/pydantic/discussions/2889
https://github.com/pydantic/pydantic/issues/8551
https://github.com/pydantic/pydantic/issues/8379
https://github.com/pydantic/pydantic/discussions/8725

and some with "resolutions" (i think incomplete ones as this is advocating for the more general case):
https://github.com/pydantic/pydantic/issues/1366
https://github.com/pydantic/pydantic/issues/892

my search query: https://github.com/pydantic/pydantic/issues?page=1&q=internal+external+field+name
